Add safe implementation of SWTBot's resetWorkbench()

briandealwis opened the pull request:
@chanseokoh noticed a sporadic failure that is due to removing the call to `SWTWorkbenchBot#resetWorkbench()` in #1348, which leaves the `DeployPropertyPageTest` dialog up and interferes with some of the subsequent tests.

```
testHelloWorld(com.google.cloud.tools.eclipse.integration.appengine.NewMavenBasedAppEngineProjectWizardTest)  Time elapsed: 31.072 sec  <<< ERROR!
org.eclipse.swtbot.swt.finder.exceptions.WidgetNotFoundException: Could not find menu bar for shell: Shell with text {Properties for foo}
	at org.eclipse.swtbot.swt.finder.SWTBotFactory.waitUntil(SWTBotFactory.java:522)
	at org.eclipse.swtbot.swt.finder.SWTBotFactory.waitUntil(SWTBotFactory.java:496)
	at org.eclipse.swtbot.swt.finder.SWTBotFactory.waitUntil(SWTBotFactory.java:484)
	at org.eclipse.swtbot.swt.finder.SWTBotFactory.waitUntilWidgetAppears(SWTBotFactory.java:466)
	at org.eclipse.swtbot.swt.finder.SWTBotFactory.menu(SWTBotFactory.java:216)
	at org.eclipse.swtbot.swt.finder.SWTBotFactory.menu(SWTBotFactory.java:204)
	at org.eclipse.swtbot.swt.finder.SWTBotFactory.menu(SWTBotFactory.java:233)
	at com.google.cloud.tools.eclipse.integration.appengine.SwtBotAppEngineActions.createMavenWebAppProject(SwtBotAppEngineActions.java:80)
```

chanseokoh commented:
Can you add the watchdog to the following tests too? I've seen timeouts for them.

`LocalAppEngineServerDelegateTest`
`AppEngineStandardProjectConvertJobTest`
`AppEngineStandardProjectConvertCommandHandlerTest`

briandealwis commented:
Sample output of the watchdog timer (similar to `jstack`)
```
+-------------------------------------------------------------------------------
| STACK DUMP @ 35.63 s: testDebugLaunch(com.google.cloud.tools.eclipse.integration.appengine.DebugNativeAppEngineStandardProjectTest)
|
| "main" Id=1 RUNNABLE
|	at org.eclipse.swt.internal.cocoa.OS.objc_msgSend_bool(Native Method)
|	at org.eclipse.swt.internal.cocoa.NSRunLoop.runMode(NSRunLoop.java:42)
|	at org.eclipse.swt.widgets.Display.sleep(Display.java:4829)
|	at org.eclipse.ui.application.WorkbenchAdvisor.eventLoopIdle(WorkbenchAdvisor.java:368)
|	at org.eclipse.ui.internal.ide.application.IDEWorkbenchAdvisor.eventLoopIdle(IDEWorkbenchAdvisor.java:972)
|	at org.eclipse.ui.internal.Workbench$1.eventLoopIdle(Workbench.java:517)
|	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$4.run(PartRenderingEngine.java:1138)
|	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
|	...
|
|
| "Reference Handler" Id=2 WAITING on java.lang.ref.Reference$Lock@69b5445d
|	at java.lang.Object.wait(Native Method)
|	-  waiting on java.lang.ref.Reference$Lock@69b5445d
|	at java.lang.Object.wait(Object.java:502)
|	at java.lang.ref.Reference.tryHandlePending(Reference.java:191)
|	at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:153)
+———————————————————————————————————————
```

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.integration.appengine/src/com/google/cloud/tools/eclipse/integration/appengine/NewNativeAppEngineStandardProjectTest.java:
```diff
@@ -29,14 +30,20 @@
 public class NewNativeAppEngineStandardProjectTest extends BaseProjectTest {
+  @Rule
+  public ThreadDumpingWatchdog timer =
+      new ThreadDumpingWatchdog(2, TimeUnit.MINUTES);
```
Now it fits in a single line.

briandealwis commented:
We seem to have another code path that tickles [bug 511729](https://bugs.eclipse.org/bugs/show_bug.cgi?id=511729).  Unfortunately the 8 frames shown is insufficient!
```
| "main" Id=1 TIMED_WAITING on org.eclipse.core.internal.jobs.Semaphore@69c37c3e
|	at java.lang.Object.wait(Native Method)
|	-  waiting on org.eclipse.core.internal.jobs.Semaphore@69c37c3e
|	at org.eclipse.core.internal.jobs.Semaphore.acquire(Semaphore.java:39)
|	at org.eclipse.core.internal.jobs.OrderedLock.doAcquire(OrderedLock.java:170)
|	at org.eclipse.core.internal.jobs.OrderedLock.acquire(OrderedLock.java:106)
|	at org.eclipse.core.internal.jobs.OrderedLock.acquire(OrderedLock.java:82)
|	at org.eclipse.wst.common.componentcore.internal.StructureEdit.getComponentModelRoot(StructureEdit.java:467)
|	at org.eclipse.wst.common.componentcore.internal.StructureEdit.getWorkbenchModules(StructureEdit.java:506)
|	at org.eclipse.wst.common.componentcore.internal.StructureEdit.getComponent(StructureEdit.java:949)
|	…
|
| "Worker-1" Id=65 BLOCKED on org.eclipse.wst.common.componentcore.internal.util.ComponentImplManager@3777fc44 owned by "main" Id=1
|	at org.eclipse.wst.common.componentcore.internal.util.ComponentImplManager.createComponent(ComponentImplManager.java:223)
|	-  blocked on org.eclipse.wst.common.componentcore.internal.util.ComponentImplManager@3777fc44
|	at org.eclipse.wst.common.componentcore.internal.util.ComponentImplManager.createComponent(ComponentImplManager.java:212)
|	at org.eclipse.wst.common.componentcore.ComponentCore.createComponent(ComponentCore.java:64)
|	at org.eclipse.wst.common.componentcore.internal.builder.DependencyGraphImpl$GraphUpdateJob$1.run(DependencyGraphImpl.java:494)
|	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
|	at org.eclipse.wst.common.componentcore.internal.builder.DependencyGraphImpl$GraphUpdateJob.run(DependencyGraphImpl.java:462)
|	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:56)
|
```

briandealwis commented:
Added an output line to indicate that the Watchdog was installed, which is useful for reconciling console output with surefire tests.
```
[Watchdog] > testDebugLaunch(com.google.cloud.tools.eclipse.integration.appengine.DebugNativeAppEngineStandardProjectTest)
```

Added a custom stack dumper, and added some support for ignoring uninteresting threads (e.g., Eclipse Jobs' worker pool, Equinox internal threads, internal JVM threads) when blocked on known types of objects.
```
+-------------------------------------------------------------------------------
| STACK DUMP @ 10.03 s: testDebugLaunch(com.google.cloud.tools.eclipse.integration.appengine.DebugNativeAppEngineStandardProjectTest)
|
| "WorkbenchTestable" [41] WAITING on java.lang.Thread@7e8a3751
|     at java.lang.Object.wait(Native Method)
|     - waiting on java.lang.Thread@7e8a3751
|     at java.lang.Thread.join(Thread.java:1249)
|     at java.lang.Thread.join(Thread.java:1323)
|     at org.eclipse.pde.internal.junit.runtime.PlatformUITestHarness.runTests(PlatformUITestHarness.java:53)
|     at org.eclipse.e4.ui.internal.workbench.swt.E4Testable.lambda$0(E4Testable.java:73)
|     at org.eclipse.e4.ui.internal.workbench.swt.E4Testable$$Lambda$100/1183302020.run(Unknown Source)
|     at java.lang.Thread.run(Thread.java:745)
| 
| "ReaderThread" [45] RUNNABLE (in native code)
|     at java.net.SocketInputStream.socketRead0(Native Method)
|     at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
|     at java.net.SocketInputStream.read(SocketInputStream.java:170)
|     at java.net.SocketInputStream.read(SocketInputStream.java:141)
|     at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
|     at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
|     at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
|     - locked java.io.InputStreamReader@7bda60f2
|     at java.io.InputStreamReader.read(InputStreamReader.java:184)
|     at java.io.BufferedReader.fill(BufferedReader.java:161)
|     at java.io.BufferedReader.readLine(BufferedReader.java:324)
|     - locked java.io.InputStreamReader@7bda60f2
|     at java.io.BufferedReader.readLine(BufferedReader.java:389)
|     at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner$ReaderThread.run(RemoteTestRunner.java:135)
| 
| Ignored threads:
|   "Reference Handler" [2] WAITING on java.lang.ref.Reference$Lock@6aa7f95e
|   "Finalizer" [3] WAITING on java.lang.ref.ReferenceQueue$Lock@7d620712
|   "Signal Dispatcher" [4] RUNNABLE
|   "JDWP Transport Listener: dt_socket" [5] RUNNABLE
|   "JDWP Event Helper Thread" [6] RUNNABLE
|   "JDWP Command Reader" [7] RUNNABLE (in native code)
|   "Worker-1" [35] TIMED_WAITING on org.eclipse.core.internal.jobs.WorkerPool@32105ec5
+-------------------------------------------------------------------------------
```

chanseokoh commented:
Just noticed this:
```
	private final boolean ignoreUselessThreads = true;
	                      ^^^^^^^^^^^^^^^^^^^^
The value of the field ThreadDumpingWatchdog.ignoreUselessThreads is not used
``

briandealwis commented:
Oops, meant to make that configurable (just in case).  Will address that with #1377.
